### PR TITLE
Async client

### PR DESCRIPTION
**Pull request by wfystx**
Async client of Graph and Meta. with examples

### CODE REVIEW

File: examples/src/main/java/com/vesoft/nebula/examples/StorageClientExample.java [hunk 1]
```diff
@@ -28,7 +26,7 @@ public static void main(String[] args) {
         }
 
         try {
-            MetaClientImpl metaClient = new MetaClientImpl(args[0], Integer.valueOf(args[1]));
+            MetaClientImpl metaClient = new MetaClientImpl(args[0], Integer.parseInt(args[1]));
```
Comment (darionyaphet) [inline examples/src/main/java/com/vesoft/nebula/examples/StorageClientExample.java]: `MetaClient`

File: client/src/main/java/com/vesoft/nebula/storage/client/async/AsyncStorageClient.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ */
+
+package com.vesoft.nebula.storage.client.async;
+
+import com.google.common.base.Optional;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.vesoft.nebula.Client;
+
+
+public interface AsyncStorageClient extends Client {
```
Comment (darionyaphet) [inline client/src/main/java/com/vesoft/nebula/storage/client/async/AsyncStorageClient.java]: If there's time, support some comment :)

File: client/src/main/java/com/vesoft/nebula/meta/client/async/AsyncMetaClientImpl.java [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+                    LOGGER.error(String.format("Get tags Error: %s", response.getCode()));
+                }
+                return Optional.absent();
+            }
+        });
+    }
+
+    @Override
+    public void close() throws Exception {
+
```
Comment (darionyaphet) [inline client/src/main/java/com/vesoft/nebula/meta/client/async/AsyncMetaClientImpl.java]: close transport at here.

File: client/src/main/java/com/vesoft/nebula/meta/client/async/AsyncMetaClientImpl.java [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+                return Optional.of(result);
+            }
+        });
+    }
+
+    @Override
+    public ListenableFuture<Optional<GetPartsAllocResult>> getPartsAlloc(int spaceId) {
+        return threadPool.submit(new Callable<Optional<GetPartsAllocResult>>() {
+            @Override
+            public Optional<GetPartsAllocResult> call() throws Exception {
```
Comment (darionyaphet) [inline client/src/main/java/com/vesoft/nebula/meta/client/async/AsyncMetaClientImpl.java]: remove `Exception `

File: client/src/main/java/com/vesoft/nebula/meta/client/MetaClient.java [hunk 5]
```diff
@@ -14,6 +14,8 @@
 
 public interface MetaClient extends Client {
 
+    public boolean listSpaces();
```
Comment (darionyaphet) [inline client/src/main/java/com/vesoft/nebula/meta/client/MetaClient.java]: why return boolean ?

File: examples/src/main/java/com/vesoft/nebula/examples/StorageClientExample.java (hunk 1 above)
Comment (darionyaphet) [inline examples/src/main/java/com/vesoft/nebula/examples/StorageClientExample.java]: Use inferface `MetaClient` would be better ?

File: examples/src/main/java/com/vesoft/nebula/examples/MetaClientExample.java [hunk 6]
```diff
@@ -14,8 +14,8 @@
     private static final Logger LOGGER = LoggerFactory.getLogger(MetaClientExample.class);
 
     public static void main(String[] args) {
-        MetaClientImpl metaClient = new MetaClientImpl("127.0.0.1", 45500);
-        LOGGER.info(metaClient.getPart(2, 1).toString());
+        MetaClientImpl metaClient = new MetaClientImpl("127.0.0.1", 28910);
```
Comment (darionyaphet) [inline examples/src/main/java/com/vesoft/nebula/examples/MetaClientExample.java]: the default port of meta service is 45500 please use the default value in examples

File: examples/src/main/java/com/vesoft/nebula/examples/AsyncGraphClientExample.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AsyncGraphClientExample {
+    private static final Logger LOGGER = LoggerFactory.getLogger(AsyncGraphClientExample.class);
+
+    private static final String SPACE_NAME = "test";
+
+    private static final String[] createTags = {
```
Comment (darionyaphet) [inline examples/src/main/java/com/vesoft/nebula/examples/AsyncGraphClientExample.java]: The sentence in example are the same?

File: client/src/main/java/com/vesoft/nebula/graph/client/GraphClientImpl.java [hunk 8]
```diff
@@ -14,11 +14,9 @@
 import com.facebook.thrift.transport.TSocket;
 import com.facebook.thrift.transport.TTransport;
 import com.facebook.thrift.transport.TTransportException;
-
```
Comment (darionyaphet) [inline client/src/main/java/com/vesoft/nebula/graph/client/GraphClientImpl.java]: don't remove the blank line

File: client/src/main/java/com/vesoft/nebula/graph/client/async/AsyncGraphClient.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+     */
+    public int connect(String username, String password);
+
+    /**
+     * Execute the DML statement.
+     *
+     * @param statement execution statement.
+     * @return
+     */
+    public ExecuteCallback execute(String statement);
```
Comment (darionyaphet) [inline client/src/main/java/com/vesoft/nebula/graph/client/async/AsyncGraphClient.java]: The query sentence is miss ?

File: client/src/main/java/com/vesoft/nebula/meta/client/MetaClientImpl.java [hunk 10]
```diff
@@ -214,11 +216,10 @@ private boolean connect() {
             try {
                 transport.open();
                 client = new MetaService.Client(protocol);
-                return true;
```
Comment (darionyaphet) [inline client/src/main/java/com/vesoft/nebula/meta/client/MetaClientImpl.java]: why remove the return value ?

File: client/src/main/java/com/vesoft/nebula/meta/client/async/AsyncMetaClient.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+     */
+    public ListSpaceCallback listSpaces();
+
+    /**
+     * Get Parts Allocations
+     *
+     * @param spaceId space ID
+     * @return
+     */
+    public GetPartsAllocCallback getPartsAlloc(int spaceId);
```
Comment (darionyaphet) [inline client/src/main/java/com/vesoft/nebula/meta/client/async/AsyncMetaClient.java]: please use space name to  replace with the space ID.

File: client/src/main/java/com/vesoft/nebula/meta/client/async/AsyncMetaClientImpl.java [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+    public AsyncMetaClientImpl(String host, int port) {
+        this(Lists.newArrayList(HostAndPort.fromParts(host, port)),
+            DEFAULT_TIMEOUT_MS, DEFAULT_CONNECTION_RETRY_SIZE);
+    }
+
+    public AsyncMetaClientImpl(List<HostAndPort> addresses) {
+        this(addresses, DEFAULT_TIMEOUT_MS, DEFAULT_CONNECTION_RETRY_SIZE);
+    }
+
+    private boolean connect() {
```
Comment (darionyaphet) [inline client/src/main/java/com/vesoft/nebula/meta/client/async/AsyncMetaClientImpl.java]: maybe I think  we should  let user call connect and don't do that in construction

File: examples/src/main/java/com/vesoft/nebula/examples/AsyncGraphClientExample.java [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+        "CREATE TAG course(name string, credits int);",
+        "CREATE TAG building(name string);",
+        "CREATE TAG student(name string, age int, gender string);",
+    };
+
+    private static final String[] createEdges = {
+        "CREATE EDGE like(likeness double);", "CREATE EDGE select(grade int);"
```
Comment (darionyaphet) [inline examples/src/main/java/com/vesoft/nebula/examples/AsyncGraphClientExample.java]: turn to a new line

File: client/pom.xml [hunk 14]
```diff
@@ -92,6 +93,7 @@
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-source-plugin</artifactId>
+                <version>3.2.0</version>
```
Comment (darionyaphet) [inline client/pom.xml]: thanks

File: client/src/main/java/com/vesoft/nebula/entry/AbstractNebulaCallback.java [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+
+    protected TBase result;
+
+    protected AtomicBoolean isReady = new AtomicBoolean(false);
+
+    public Optional<?> getResult() throws InterruptedException {
+        while (!checkReady()) {
+            Thread.sleep(100);
+        }
+        return result == null ? Optional.absent() : Optional.of(result);
```
Comment (darionyaphet) [inline client/src/main/java/com/vesoft/nebula/entry/AbstractNebulaCallback.java]: We can make sure the result is not empty, so maybe return `Optional.of(result)` is OK .

File: client/src/main/java/com/vesoft/nebula/entry/AbstractNebulaCallback.java [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+            isReady.compareAndSet(false, true);
+        } catch (TException e) {
+            e.printStackTrace();
+        }
+    }
+
+    public abstract void doComplete(TAsyncMethodCall response) throws TException;
+
+    public boolean checkReady() {
+        if (isReady.compareAndSet(true, true)) {
```
Comment (darionyaphet) [inline client/src/main/java/com/vesoft/nebula/entry/AbstractNebulaCallback.java]: isReady.get() ?

File: client/src/main/java/com/vesoft/nebula/entry/AbstractNebulaCallback.java (hunk 15 above)
Comment (wfystx) [reply client/src/main/java/com/vesoft/nebula/entry/AbstractNebulaCallback.java]: 👍

File: client/src/main/java/com/vesoft/nebula/entry/AbstractNebulaCallback.java (hunk 16 above)
Comment (wfystx) [reply client/src/main/java/com/vesoft/nebula/entry/AbstractNebulaCallback.java]: 👌

**Pull request merged**